File: frontend/src/services/productService.ts
import { api } from "../api/axios";
import type { Product } from "../types/product";
import type { ProductQuery } from "../types/productQuery";

export async function getProducts(query?: ProductQuery): Promise<Product[]> {
  const params = new URLSearchParams();

  if (query) {
    if (query.search) params.append("Search", query.search);
    if (query.minPrice !== undefined) params.append("MinPrice", query.minPrice.toString());
    if (query.maxPrice !== undefined) params.append("MaxPrice", query.maxPrice.toString());
    if (query.page !== undefined) params.append("Page", query.page.toString());
    if (query.pageSize !== undefined) params.append("PageSize", query.pageSize.toString());
    if (query.categoryIds) {
      query.categoryIds.forEach((id) => params.append("CategoryIds", id.toString()));
    }
  }

  const response = await api.get<Product[]>(`/products?${params.toString()}`);

  return response.data;
}

export async function getProductById(id: number): Promise<Product> {
  const response = await api.get<Product>(`/products/${id}`);

  return response.data;
}

export async function createProduct(product: {
  name: string;
  description: string;
  price: number;
  stock: number;
  imageUrl: string;
  categoryId: number;
}) {
  const token = localStorage.getItem("token");

  const response = await api.post("/products", product, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  return response.data;
}

export async function updateProduct(
  id: number,
  product: {
    name: string;
    description: string;
    price: number;
    stock: number;
    imageUrl: string;
    categoryId: number;
  },
) {
  const token = localStorage.getItem("token");

  const response = await api.put(`/products/${id}`, product, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  return response.data;
}

export async function deleteProduct(id: number) {
  const token = localStorage.getItem("token");

  return api.delete(`/products/${id}`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });
}

File: backend/ECommerce.API/Services/Implementations/ProductService.cs
using ECommerce.API.Data;
using ECommerce.API.DTOs.Requests;
using ECommerce.API.DTOs.Responses;
using ECommerce.API.Models;
using ECommerce.API.Repositories.Interfaces;
using ECommerce.API.Services.Interfaces;
using AutoMapper;
using Microsoft.EntityFrameworkCore;

namespace ECommerce.API.Services.Implementations;

public class ProductService : IProductService
{
    private readonly IGenericRepository<Product> _productRepository;
    private readonly IMapper _mapper;
    private readonly IGenericRepository<Category> _categoryRepository;
    private readonly ApplicationDbContext _context;

    public ProductService(
    IGenericRepository<Product> productRepository,
    IGenericRepository<Category> categoryRepository,
    ApplicationDbContext context,
    IMapper mapper)
    {
        _productRepository = productRepository;
        _categoryRepository = categoryRepository;
        _context = context;
        _mapper = mapper;
    }
    public async Task<IEnumerable<ProductResponse>> GetAllAsync()
    {
        var products = await _productRepository.GetAllAsync();

        return _mapper.Map<IEnumerable<ProductResponse>>(products);
    }

    public async Task<IEnumerable<ProductResponse>>
        GetFilteredAsync(ProductQueryRequest request)
    {
        var query = _context.Products
            .Include(p => p.Seller)
            .Include(p => p.Category)
            .AsQueryable();

        if (!string.IsNullOrWhiteSpace(request.Search))
        {
            var searchTerm = request.Search.ToLower();
            query = query.Where(x =>
                x.Name.ToLower().Contains(searchTerm) || 
                (x.Category != null && x.Category.Name.ToLower().Contains(searchTerm)));
        }

        if (request.CategoryIds != null &&
    request.CategoryIds.Any())
        {
            query = query.Where(x =>
                request.CategoryIds.Contains(x.CategoryId));
        }

        if (request.MinPrice.HasValue)
        {
            query = query.Where(x =>
                x.Price >= request.MinPrice.Value);
        }

        if (request.MaxPrice.HasValue)
        {
            query = query.Where(x =>
                x.Price <= request.MaxPrice.Value);
        }

        query = query
            .Skip((request.Page - 1) * request.PageSize)
            .Take(request.PageSize);

        var products = await query.ToListAsync();

        return _mapper.Map<IEnumerable<ProductResponse>>(products);
    }

    public async Task<ProductResponse?> GetByIdAsync(int id)
    {
        var product = await _context.Products
            .Include(p => p.Seller)
            .FirstOrDefaultAsync(p => p.Id == id);

        if (product is null)
            return null;

        return _mapper.Map<ProductResponse>(product);
    }

    public async Task<ProductResponse> CreateAsync(
        CreateProductRequest request)
    {
        var category = await _categoryRepository
            .GetByIdAsync(request.CategoryId);

        if (category is null)
            throw new Exception("Category not found.");

        var product = new Product
        {
            Name = request.Name,
            Description = request.Description,
            Price = request.Price,
            Stock = request.Stock,
            ImageUrl = request.ImageUrl,
            CategoryId = request.CategoryId,
            CompareAtPrice = request.CompareAtPrice,
            InstallmentPlan = request.InstallmentPlan,
            Images = request.Images,
            Brand = request.Brand,
            IsActive = request.IsActive,
            SKU = request.SKU,
            TrackQuantity = request.TrackQuantity,
            ContinueSellingWhenOutOfStock = request.ContinueSellingWhenOutOfStock,
            UrlHandle = request.UrlHandle,
            MetaDescription = request.MetaDescription,
            ProductType = request.ProductType,
            Vendor = request.Vendor,
            Tags = request.Tags,
            Variants = request.Variants,
            Specifications = request.Specifications,
            Features = request.Features
        };

        await _productRepository.AddAsync(product);

        return _mapper.Map<ProductResponse>(product);
    }

    public async Task<ProductResponse?> UpdateAsync(
        int id,
        UpdateProductRequest request)
    {
        var product = await _productRepository.GetByIdAsync(id);

        if (product is null)
            return null;

        product.Name = request.Name;
        product.Description = request.Description;
        product.Price = request.Price;
        product.Stock = request.Stock;
        product.ImageUrl = request.ImageUrl;
        product.CategoryId = request.CategoryId;
        product.CompareAtPrice = request.CompareAtPrice;
        product.InstallmentPlan = request.InstallmentPlan;
        product.Images = request.Images;
        product.Brand = request.Brand;
        product.IsActive = request.IsActive;
        product.SKU = request.SKU;
        product.TrackQuantity = request.TrackQuantity;
        product.ContinueSellingWhenOutOfStock = request.ContinueSellingWhenOutOfStock;
        product.UrlHandle = request.UrlHandle;
        product.MetaDescription = request.MetaDescription;
        product.ProductType = request.ProductType;
        product.Vendor = request.Vendor;
        product.Tags = request.Tags;
        product.Variants = request.Variants;
        product.Specifications = request.Specifications;
        product.Features = request.Features;

        await _productRepository.UpdateAsync(product);

        return _mapper.Map<ProductResponse>(product);
    }

    public async Task<bool> DeleteAsync(int id)
    {
        var product = await _productRepository.GetByIdAsync(id);

        if (product is null)
            return false;

        await _productRepository.DeleteAsync(product);

        return true;
    }
}